src/atomic.rs

```diff
@@ -0,0 +1,160 @@
+use cid::multihash::Code::Blake2b256;
+use cid::multihash::MultihashDigest;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use fvm_ipld_encoding::tuple::Deserialize_tuple;
+use fvm_ipld_encoding::{serde_bytes, Cbor, RawBytes, DAG_CBOR};
+use fvm_shared::MethodNum;
+use primitives::{TCid, THamt};
+use serde::{de::DeserializeOwned, Serialize};
+
+/// Trait that determines the functions that need to be implemented by
+/// a state object to be lockable and be used in an atomic execution.
+///
+/// Different strategies may be used to merge different locked state to
+/// prepare the actor state for the execution, and for the merging of the
+/// output of the execution to the original state of the actor.
+pub trait MergeableState<S: Serialize + DeserializeOwned> {
+    /// Merge a locked state (not necessarily the output) to the current state.
+    fn merge(&mut self, other: Self) -> anyhow::Result<()>;
+    /// Merge the output of an execution to the current state.
+    fn merge_output(&mut self, other: Self) -> anyhow::Result<()>;
+}
+
+/// Internal map kept by actor supporting atomic executions to track
+/// the states that have been locked and are used in an atomic exec.
+#[allow(dead_code)]
+pub type LockedMap<T> = TCid<THamt<Cid, LockableState<T>>>;
+
+/// Trait that specifies the interface of an actor state able to support
+/// atomic executions.
+pub trait LockableActorState<T>
+where
+    T: Serialize + DeserializeOwned + MergeableState<T>,
+{
+    /// Map with all the locked state in the actor uniquely identified through
+    /// their Cid.
+    fn locked_map_cid(&self) -> LockedMap<T>;
+    /// Returns the output state of an execution from the current state
+    /// of the actor according to the input parameters.
+    fn output(&self, params: LockParams) -> LockableState<T>;
+}
+
+/// Return type for all actor functions.
+///
+/// It returns an option for developers to optionally choose if
+/// to return an output in the function.
+#[allow(dead_code)]
+type ActorResult = anyhow::Result<Option<RawBytes>>;
+
+/// Trait for an actor able to support an atomic execution.
+///
+/// The functions of this trait represent the set of methods that
+/// and actor support atomic executions needs to implement. Correspondingly,
+/// it follows the same return convention used for every FVM actor method.
+pub trait LockableActor<T, S>
```

Comment:
I don't think these functions match what the coordinator expects, right? I am having a hard time to follow the end-to-end of an execution and the interactions between the atomic-exec actors and the coordinator.

Reply:
Right, I'm going to tackle this next.

Reply:
Gotcha. Resolving this in the meantime then.